### ISSUE TITLE
The variable selected_book is used in bookinstance_form.pug but it is not defined

### ISSUE THREAD

**Issue by larhou** [mdn/content#27956]
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/forms/Create_BookInstance_form

### What specific section or headline is this issue about?

View

### What information was incorrect, unhelpful, or incomplete?

        for book in book_list
          option(value=book._id, selected=(selected_book==book._id.toString() ? 'selected' : false) ) #{book.title}

The `selected_book` variable is not defined anywhere.

It is not defined in the bookinstance_form.pug file.

It is also not defined in the controller from where the variable normally would be passed:

```
// Display BookInstance create form on GET.
exports.bookinstance_create_get = asyncHandler(async (req, res, next) => {
  const allBooks = await Book.find({}, "title").exec();

  res.render("bookinstance_form", {
    title: "Create BookInstance",
    book_list: allBooks,
  });
});

```

### What did you expect to see?

Not sure what is the best solution.

```
        if bookinstance
          option(value=book._id selected=(bookinstance.book.toString()==book._id.toString() ? 'selected' : false)) #{book.title}
        else
          option(value=book._id) #{book.title}
```

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/learn/server-side/express_nodejs/forms/create_bookinstance_form`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/forms/Create_BookInstance_form
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/server-side/express_nodejs/forms/create_bookinstance_form/index.md
* Last commit: https://github.com/mdn/content/commit/bb026bcb88b7f45374d602301b7b0db5a49ff303
* Document last modified: 2023-07-03T10:12:35.000Z

</details>

**Comment by hamishwillee** [comment]
It's defined in the controller when the selected book can possibly have been selected. In `bookinstance_create_get` it cannot have been selected because in this case you're displaying the form first time with no data.
The pug works for this case - there is no entered value, so it will be `undefined`

It is defined in the controller when the value might already have been set. In practise this happens in the post route if there is an error. See  `bookinstance_create_post` in  https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/forms/Create_BookInstance_form#controller%E2%80%94post_route

Closing, because I believe this to be invalid. Happy to discuss if you think I have missed something.

**Comment by larhou** [comment]
Hi Hamish

Thanks !

Yep, I fully understand the logic. Selected_book is not needed for displaying the form the first time. 

However, I thought that a variable that is not defined would throw an error in Javascript.

I did a console.log(typeof selected_book === 'undefined') in bookinstanceController and it indeed resolved to true.

I think for us beginners it can be confusing that selected_book is not defined. 

This could maybe be easier to understand :

 res.render("bookinstance_form", {
   title: "Create BookInstance",
   book_list: allBooks,
   selected_book: undefined,
 });

😉

BTW: I simply love this tutorial. Amazingly well done and structured. It has helped me tremendously. I am dreaming/hoping that one day it can be expanded with user registration and authentication with Express sessions. 

Lars

**Comment by hamishwillee** [comment]
Hi @larhou ,

Glad you're enjoying the tutorial. [undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined) is a primitive type in JavaScript - it is a common and expected type just meaning "has not been assigned a value". As will any other type, you will get errors if you misuse it - for example by calling a function on it that does not exist.

A programmer would not usually explicitly pass `undefined` as suggested. For the same reason we don't pass in `errors` for this case. 

I've therefore addressed your concern with words - see #28016

**Comment by hamishwillee** [comment]
PS It is unlikely that the tutorial will be extended any time soon. Just maintaining it is quite time consuming.
